### DISCUSSION TITLE
Is it safe to use async event handler?

### DISCUSSION THREAD

**Discussion by ultimaweapon** [gircore/gir.core#853] [Q&A] [answered]
I was tried to find the implementation of `SynchronizationContext` in this repository but not found so I'm not sure if it safe to use async on the event handler.

**Comment by badcel** [comment]
Can you elaborate a bit more on the SychronizatiomContext? (My knowledge on this topic is quite limited.)
What is your use case? Did you experience problems with the current implementation? What do you expect to be available on the current codebase? Can you give a sample?

In General I read through the [documentation](https://learn.microsoft.com/en-us/dotnet/api/system.threading.synchronizationcontext?view=net-8.0) and can confirm that there is currently no GirCore specific implementation of the SynchronizationContext.

It reads like it is needed to make async/await work without the need to manually dispatch the code back to the main loop to alter UI elements. Can you confirm this?

Additionally we have currently issue #845 which could be related. Do you know if the synchronization context could help to catch those unhandled exceptions to avoid crashing the runtime?

Having some samples would be really great as those could be used to create unit tests to ensure the desired behavior.

**Reply by fsobolev** [reply]
I don't know if this information will be useful since I guess it doesn't answer any of your questions really, but this is related to async calls so I hope it will help somehow.
Recently @nlogozzo and me were fighting GTK crashes in Denaro and Tube Converter, and these crashes were caused by GTK calls from another thread. That's what we learned:
1. GTK calls that change properties seems to be mostly safe to use on another thread (for example, changing the text of a `Gtk.Label`), but anything that changes the layout can cause crash
2. These crashes are random and/or depend on hardware. Nick didn't experience crashes at all, I only experienced them in Tube Converter and it was random, but some users both for Denaro and Tube Converter reported about constant crashes (https://github.com/NickvisionApps/Denaro/issues/401 https://github.com/NickvisionApps/TubeConverter/issues/130)
3. These crashes are either silent or give `malloc` error. Using try-catch doesn't help to get any additional info about the crash.

In Denaro and Tube Converter we don't use SynchronizationContext, because there are a lot of places where we not only use GTK calls after `await` in methods that were initially called from UI thread (this is where SyncronizationContext could help, and this is what happens in WinUI because context is automatically set there), but also use GTK calls from methods running on background thread, so we just use `g_main_context_invoke` everywhere, it works flawlessly (example can be found [here](https://github.com/NickvisionApps/Denaro/blob/main/NickvisionMoney.GNOME/Controls/TransactionRow.cs)).

**Reply by ultimaweapon** [reply]
Given the following code:

```cs
var button = Button.NewWithLabel("Action");

button.OnClicked += async (sender, e) =>
{
    await db.ReadAsync();
};
```

AFAIK the GTK required every calls into the widget to be from the main thread. That mean I can't touch any widgets after the `await` if the current `SynchronizationContext` is not scheduled the continuation back to the same thread before the `await`.

> It reads like it is needed to make async/await work without the need to manually dispatch the code back to the main loop to alter UI elements. Can you confirm this?

Right. `SynchronizationContext` used to control what thread to execute the code after the `await` that does not have `.ConfigureAwait(false)`.

> Additionally we have currently issue https://github.com/gircore/gir.core/issues/845 which could be related. Do you know if the synchronization context could help to catch those unhandled exceptions to avoid crashing the runtime?

I don't think `SynchronizationContext` will solve #845 because all it does is resume the code after the `await` in the correct thread.

I'm happy to help with the unit tests so let me know if you are okay with the PR for the unit test only. I want to help with the implementation of `SynchronizationContext` but I already take a look on the implementation on https://github.com/GtkSharp/GtkSharp, which is LGPL and I believe I'm not eligible to implement this code for MIT license.

**Reply by badcel** [reply]
Right an implementation must be a new one.

If you are willing to help it would be great. I will make this discussion an issue.

I would assume it is needed to start some native main loop (probably from the GirTest project) and then create a native event which is subscribed to in C#. Somehow the test would then need to detect that the actions in the test are executed on the right thread.

see #854

**Reply by cameronwhite** [reply]
I've also previously looked at the original code, and while GtkSharp has an LGPL license this specific code appears to be MIT licensed?
`GLibSynchronizationContext.cs` from GtkSharp has this license header which is MIT
```
//
// GLibSynchronizationContext.cs
// 
// Author:
//       Rickard Edström <ickard@gmail.com>
//
// Copyright (c) 2010 Rickard Edström
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
```

**Reply by ultimaweapon** [reply]
@cameronwhite you are right. @badcel if you are okay with this I can implement both `SynchronizationContext` and the unit tests in a single PR. If not I'll add only unit tests.

**Reply by badcel** [reply]
@cameronwhite it looks like it. 

@ultimaweapon Nevertheless please don't copy the code but seriously work on this topic. Test things out, make yourself familiar with the scenario, understand what happens and what needs to happen. If you do it like this it is your own work and it has the chance on being better than the one from GtkSharp.

Another question would be where to put the new class into action.

**Reply by ultimaweapon** [reply]
Maybe inside `src/Libs/GLib-2.0/Public` because we need to use https://docs.gtk.org/glib/func.idle_add.html? Then we can set it as a default `SynchronizationContext` inside `GLib.Initialize` method.

**Reply by badcel** [reply]
Is there need to have the `SynchronizationContext` available in the public namespace? `GLib.Module.Initialize` could call into the `GLib.Internal` namespace, too. Calling the `GLib` namespace from `GLib.Internal` is not a blocker for me.

I would consider it to be needed in the `GLib` namespace if the user is expected to use the class themself. I never used the synchronization context myself. Is there some usecase to use it directly? If it is just there to "make things work" it could live in the `Internal` namespace for now.

With version 1.0 there will be some API stability guarantees. If we move the class into `GLib` namespace it would need to keep it's API otherwise it would be a major version change. I'm not yet sure if the API will be guaranteed to be stable in the `Internal` namespace or if it is like "use on your own risk" which would allow internal refactorings and optimizations without doing a major version jump.

**Reply by ultimaweapon** [reply]
Okay I agree to put it in `GLib.Internal` because it is unusual for people to use `SynchronizationContext` directly.

**Comment by ultimaweapon** [comment, marked as answer]
The PR has arrived #856!
